packages/parse5-parser-stream/lib/index.ts

```diff
@@ -0,0 +1,105 @@
+import { Writable } from 'node:stream';
+import { Parser, ParserOptions } from 'parse5/dist/parser/index.js';
+import type { TreeAdapterTypeMap } from 'parse5/dist/tree-adapters/interface.js';
+import type { DefaultTreeAdapterMap } from 'parse5/dist/tree-adapters/default.js';
+
+/**
+ * Streaming HTML parser with scripting support.
+ * A [writable stream](https://nodejs.org/api/stream.html#stream_class_stream_writable).
+ *
+ * @example
+ *
+ * ```js
+ * const ParserStream = require('parse5-parser-stream');
+ * const http = require('http');
+ *
+ * // Fetch the page content and obtain it's <head> node
+ * http.get('http://inikulin.github.io/parse5/', res => {
+ *     const parser = new ParserStream();
+ *
+ *     parser.once('finish', () => {
+ *         console.log(parser.document.childNodes[1].childNodes[0].tagName); //> 'head'
+ *     });
+ *
+ *     res.pipe(parser);
+ * });
+ * ```
+ *
+ */
+export class ParserStream<T extends TreeAdapterTypeMap = DefaultTreeAdapterMap> extends Writable {
+    private lastChunkWritten = false;
+    private writeCallback: null | (() => void) = null;
+    private pausedByScript = false;
+
+    public parser: Parser<T>;
+    private pendingHtmlInsertions: string[] = [];
+    /** The resulting document node. */
+    public document: T['document'];
+
+    constructor(options?: ParserOptions<T>) {
+        super({ decodeStrings: false });
+
+        this.parser = new Parser(options);
+
+        this.document = this.parser.treeAdapter.createDocument();
+        this.parser._bootstrap(this.document, null);
+    }
+
+    //WritableStream implementation
+    override _write(chunk: string, _encoding: string, callback: () => void): void {
+        if (typeof chunk !== 'string') {
+            throw new TypeError('Parser can work only with string streams.');
+        }
+
+        this.writeCallback = callback;
+        this.parser.tokenizer.write(chunk, this.lastChunkWritten);
+        this._runParsingLoop();
+    }
+
+    // TODO [engine:node@>=16]: Due to issues with Node < 16, we are overriding `end` instead of `_final`.
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    override end(chunk?: any, encoding?: any, callback?: any): any {
```

Comment:
it would be good to drop the `any`, its a shame this one of node's typical overloaded methods, but it could be this i think:

```ts
override end(cb?: () => void): Writable;
override end(chunk?: any, cb?: () => void): Writable;
override end(
  chunk?: any, // not sure we can get rid, the DT types need fixing so this is `unknown`
  encoding: BufferEncoding,
  cb?: () => void
): Writable;
```

the base one has those 3 overloads. i suppose these all being any is why the compiler allowed it

Reply:
Agreed that the code is currently super hacky. In Node < 16, trailing data would not be parsed properly, which is made me revert my previous beautiful `_final` override. The code is now equivalent to what was there before.

The [expiring comment](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/expiring-todo-comments.md) above will start failing once failing Node versions are no longer supported, which is why I'd leave things the way they are for now.